src/backend/mpi/swdsm.cpp

```diff
@@ -150,70 +151,62 @@ unsigned long isPowerOf2(unsigned long x){
   return retval;
 }
 
-void flushWriteBuffer(void){
-	unsigned long i,j;
-	double t1,t2;
-
-	t1 = MPI_Wtime();
+void flush_write_buffer(){
+	double t1 = MPI_Wtime();
 	pthread_mutex_lock(&wbmutex);
 
-  for(i = 0; i < cachesize; i+=CACHELINE){
-    unsigned long distrAddr = cacheControl[i].tag;
-    if(distrAddr != GLOBAL_NULL){
-
-      unsigned long distrAddr = cacheControl[i].tag;
-      unsigned long lineAddr = distrAddr/(CACHELINE*pagesize);
-      lineAddr*=(pagesize*CACHELINE);
-			void * lineptr = (char*)startAddr + lineAddr;
+	/* Sort the buffer to ensure that MPI windows are locked in order */
+	if(!write_buffer->empty()){
+		write_buffer->sort();
+	}
 
-      argo_byte dirty = cacheControl[i].dirty;
-			if(dirty == DIRTY){
-				mprotect(lineptr, pagesize*CACHELINE, PROT_READ);
-				cacheControl[i].dirty=CLEAN;
-				for(j=0; j < CACHELINE; j++){
-					storepageDIFF(i+j,pagesize*j+lineAddr);
-				}
-			}
-    }
-  }
+	/* Empty the write buffer */
+	while(!write_buffer->empty()){
+		unsigned long cache_index = write_buffer->pop();
+		unsigned long page_address = cacheControl[cache_index].tag;
+		void* page_ptr = static_cast<char*>(startAddr) + page_address;
+
+		/* Since the page is guaranteed to be dirty, write it back */
+		mprotect(page_ptr, pagesize*CACHELINE, PROT_READ);
+		cacheControl[cache_index].dirty=CLEAN;
+		for(int i=0; i < CACHELINE; i++){
+			storepageDIFF(cache_index+i,pagesize*i+page_address);
+		}
+	}
 
-	for(i = 0; i < (unsigned long)numtasks; i++){
+	/* Close any windows used to write back data */
+	for(int i = 0; i < numtasks; i++){
 		if(barwindowsused[i] == 1){
 			MPI_Win_unlock(i, globalDataWindow[i]);
 			barwindowsused[i] = 0;
 		}
 	}
 
-	writebufferstart = 0;
-	writebufferend = 0;
-
 	pthread_mutex_unlock(&wbmutex);
-	t2 = MPI_Wtime();
+	double t2 = MPI_Wtime();
 	stats.flushtime += t2-t1;
 }
 
-void addToWriteBuffer(unsigned long cacheIndex){
+void add_to_write_buffer(unsigned long cache_index){
```

Comment:
Shouldn't this be a member of the WriteBuffer class?

Reply:
flush_write_buffer is now WriteBuffer::flush().
write_back_write_buffer is now WriteBuffer::flush_partial()
add_to_write_buffer is now WriteBuffer::add(T val)
WriteBuffer::erase(T val) is used to erase a single element.